File: backend/app/api/schemas.py
"""Pydantic response models — the stable contract consumed by the frontend."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel


class HealthResponse(BaseModel):
    status: str = "ok"
    version: str
    env: str


class PairSummary(BaseModel):
    name: str
    status: str                      # pending | clean | defect | error
    total_defects: int = 0
    defect: int = 0
    review: int = 0
    quality_score: Optional[int] = None  # category-weighted /100
    error: Optional[str] = None
    report_url: Optional[str] = None  # present once an HTML report exists


class CompareAccepted(BaseModel):
    job_id: str
    status: str                      # queued
    pair_count: int
    unmatched_reference: list[str] = []
    unmatched_candidate: list[str] = []


class JobResponse(BaseModel):
    job_id: str
    status: str                      # queued | running | completed | failed
    created_at: float
    updated_at: float
    pairs: list[PairSummary] = []
    unmatched_reference: list[str] = []
    unmatched_candidate: list[str] = []
    error: Optional[str] = None


class ErrorResponse(BaseModel):
    error: str
    detail: Optional[str] = None
    request_id: Optional[str] = None

File: frontend/src/lib/schemas.ts
import { z } from 'zod';

export const pairStatusSchema = z.enum(['pending', 'clean', 'defect', 'error']);

export const pairSummarySchema = z.object({
  name: z.string(),
  status: pairStatusSchema,
  total_defects: z.number().default(0),
  defect: z.number().default(0),
  review: z.number().default(0),
  quality_score: z.number().nullable().optional().default(null),
  error: z.string().nullable().optional().default(null),
  report_url: z.string().nullable().optional().default(null),
});

export const compareAcceptedSchema = z.object({
  job_id: z.string(),
  status: z.string(),
  pair_count: z.number(),
  unmatched_reference: z.array(z.string()).default([]),
  unmatched_candidate: z.array(z.string()).default([]),
});

export const jobStatusSchema = z.enum(['queued', 'running', 'completed', 'failed']);

export const jobResponseSchema = z.object({
  job_id: z.string(),
  status: jobStatusSchema,
  created_at: z.number(),
  updated_at: z.number(),
  pairs: z.array(pairSummarySchema).default([]),
  unmatched_reference: z.array(z.string()).default([]),
  unmatched_candidate: z.array(z.string()).default([]),
  error: z.string().nullable().optional().default(null),
});

export const errorResponseSchema = z.object({
  error: z.string(),
  detail: z.string().nullable().optional().default(null),
  request_id: z.string().nullable().optional().default(null),
});
